**Please visit https://ProStepSky.github.io/.**

This work © 2024 by "ProStepSky.github.io จัดทำโดย S.U." is licensed under [CC BY-NC-ND 4.0](http://creativecommons.org/licenses/by-nc-nd/4.0/)  
(อ้างอิงแหล่งที่มา ห้ามนำไปใช้เพื่อการค้า และห้ามดัดแปลง)

---

# Import Libraries

In [1]:
import numpy as np
import pandas as pd

In [2]:
# Suppress all warnings 
import warnings
warnings.filterwarnings("ignore")

# Import Data

In [3]:
loan_data_preprocessed_backup = pd.read_csv('./data/loan_data_2007_2014_preprocessed.csv')

# Explore Data

In [4]:
loan_data_preprocessed = loan_data_preprocessed_backup.copy()

In [5]:
loan_data_preprocessed.columns.values

array(['Unnamed: 0.1', 'Unnamed: 0', 'id', 'member_id', 'loan_amnt',
       'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate',
       'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
       'loan_status', 'pymnt_plan', 'url', 'desc', 'purpose', 'title',
       'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line',
       'inq_last_6mths', 'mths_since_last_delinq',
       'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal',
       'revol_util', 'total_acc', 'initial_list_status', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv',
       'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee',
       'recoveries', 'collection_recovery_fee', 'last_pymnt_d',
       'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'policy_code', 'application_type', 'annual_inc_joint', 'dti_join

In [6]:
loan_data_preprocessed.head()

Unnamed: 0.1  Unnamed: 0       id  member_id  loan_amnt  funded_amnt  \
0             0           0  1077501    1296599       5000         5000   
1             1           1  1077430    1314167       2500         2500   
2             2           2  1077175    1313524       2400         2400   
3             3           3  1076863    1277178      10000        10000   
4             4           4  1075358    1311748       3000         3000   

   funded_amnt_inv        term  int_rate  installment  ... addr_state:UT  \
0           4975.0   36 months     10.65       162.87  ...         False   
1           2500.0   60 months     15.27        59.83  ...         False   
2           2400.0   36 months     15.96        84.33  ...         False   
3          10000.0   36 months     13.49       339.31  ...         False   
4           3000.0   60 months     12.69        67.79  ...         False   

  addr_state:VA addr_state:VT addr_state:WA addr_state:WI  addr_state:WV  \
0         False         False         False         False          False   
1         False         False         False         False          False   
2         False         False         False         False          False   
3         False         False         False         False          False   
4         False         False         False         False          False   

  addr_state:WY initial_list_status:f initial_list_status:w good_bad  
0         False                  True                 False        1  
1         False                  True                 False        0  
2         False                  True                 False        1  
3         False                  True                 False        1  
4         False                  True                 False        1  

[5 rows x 209 columns]

In [7]:
loan_data_preprocessed.tail()

Unnamed: 0.1  Unnamed: 0       id  member_id  loan_amnt  funded_amnt  \
466280        466280      466280  8598660    1440975      18400        18400   
466281        466281      466281  9684700   11536848      22000        22000   
466282        466282      466282  9584776   11436914      20700        20700   
466283        466283      466283  9604874   11457002       2000         2000   
466284        466284      466284  9199665   11061576      10000        10000   

        funded_amnt_inv        term  int_rate  installment  ... addr_state:UT  \
466280          18400.0   60 months     14.47       432.64  ...         False   
466281          22000.0   60 months     19.97       582.50  ...         False   
466282          20700.0   60 months     16.99       514.34  ...         False   
466283           2000.0   36 months      7.90        62.59  ...         False   
466284           9975.0   36 months     19.20       367.58  ...         False   

       addr_state:VA addr_state:VT addr_state:WA addr_state:WI  addr_state:WV  \
466280         False         False         False         False          False   
466281         False         False         False         False          False   
466282         False         False         False         False          False   
466283         False         False         False         False          False   
466284         False         False         False         False          False   

       addr_state:WY initial_list_status:f initial_list_status:w good_bad  
466280         False                 False                  True        1  
466281         False                  True                 False        0  
466282         False                  True                 False        1  
466283         False                 False                  True        1  
466284         False                  True                 False        1  

[5 rows x 209 columns]

ปกติแล้ว LGD และ EAD จะเกิดขึ้นเมื่อผู้ขอกู้ยืมได้ผิดนัดชำระหนี้แล้วเท่านั้น

- Loss Given Default (LGD): เปอร์เซนต์ความเสียหายเมื่อลูกหนี้ผิดนัดชำระหนี้

- Exposure at Default (EAD): ยอดหนี้เมื่อลูกหนี้ผิดชำระหนี้

หากนำการกู้ยืมที่ผิดนัดชำระหนี้มาใช้สร้าง LGD หรือ EAD models เลย อาจจะไม่ค่อยถูกต้องนัก เพราะเมื่อผู้ขอกู้ได้ผิดนัดชำระหนี้แล้ว ผู้ขอกู้คนนั้นอาจจะชำระหนี้ไปบางส่วนแล้วก็ได้ (ความเสียหายส่วนนั้นก็จะได้รับการ recovered) ดังนั้น แนวปฏิบัติที่ดีคือควรเว้นระยะเวลาให้เพียงพอสำหรับการชำระหนี้ล่าช้า

ดังนั้น จะคัดเลือกเฉพาะการกู้ยืมที่มีสถานะ loan_status เป็น "Charged Off" และ "Does not meet the credit policy. Status:Charged Off" เท่านั้น (สถานะเป็น "Default" และ "Late (31-120 days)" ยังไม่ถือว่าเป็นหนี้สูญ (written-off))

ทำการคัดเลือกเฉพาะการกู้ยืมที่บันทึกสถานะไว้ว่าเป็นหนี้สูญ (written-off):

In [8]:
loan_data_defaults = loan_data_preprocessed[loan_data_preprocessed['loan_status'].isin(['Charged Off',
                                                                                        'Does not meet the credit policy. Status:Charged Off'])]

In [9]:
loan_data_defaults.shape

(43236, 209)

พบผู้ขอกู้ที่ผิดนัดชำระหนี้จำนวน 43236 คน

ตรวจสอบดูว่ายังมี missing values เหลืออยู่ไหม:

In [10]:
pd.options.display.max_rows = None
# Sets the pandas DataFrame options to display all columns/ rows.

In [11]:
loan_data_defaults.isnull().sum()

Unnamed: 0.1                                                           0
Unnamed: 0                                                             0
id                                                                     0
member_id                                                              0
loan_amnt                                                              0
funded_amnt                                                            0
funded_amnt_inv                                                        0
term                                                                   0
int_rate                                                               0
installment                                                            0
grade                                                                  0
sub_grade                                                              0
emp_title                                                           3287
emp_length                                         

มีตัวแปรหลายตัวที่มี missing values แต่ตัวแปรที่มี missing values และจะนำมาใช้สร้าง model มีแค่ `mths_since_last_delinq` และ `mths_since_last_record`

ตอนที่เคยสร้าง PD model นั้น ตัวแปรอิสระจำเป็นต้องเป็น dummy variables เท่านั้น ดังนั้น จึงรวม missing values ให้เป็น 1 dummy variable

แต่ตอนนี้ การสร้าง LGD และ EAD models นั้น ไม่จำเป็นต้องทำให้ผู้คนที่มาจากทุก background สามารถเข้าใจตัว model ได้ง่าย เหมือนตอนที่เคยสร้าง PD model อีกแล้ว เพราะ LGD และ EAD models นี้ มีแต่ risk modellers เท่านั้นที่ใช้ ดังนั้น ตัวแปรอิสระใน LGD และ EAD models ไม่จำเป็นต้องเป็น dummy variables เท่านั้นอีกแล้ว

# Independent Variables

ในการสร้าง LGD และ EAD models จะใช้ตัวแปรอิสระตัวเดียวกับที่เคยใช้ในการสร้าง PD model แต่การสร้าง LGD และ EAD models ครั้งนี้ จะไม่สนใจ Weight of Evidence (WoE) หรือหลักเกณฑ์อื่น ๆ ที่ใช้ในการแบ่ง dummy variables อีกแล้ว

- สำหรับ discrete variables: จะทำการสร้าง dummy variables ให้มีจำนวนเท่ากันกับจำนวน categories ที่มี

- สำหรับ continuous variables: จะใช้ตามเดิมเลยก็ได้หรือจะ transform ก็ได้ และไม่จำเป็นต้องทำ fine classing หรือ coarse classing

สำหรับตัวแปร `mths_since_last_delinq` (จำนวนเดือนที่ผ่านไปนับจากการขาดการชำระเงินล่าสุด) และ `mths_since_last_record` (จำนวนเดือนที่ผ่านไปนับจากที่มีการบันทึกประวัติเสียทางการเงินล่าสุด) ในการสร้าง LGD และ EAD models นี้ จะนำมาใช้ในแบบ continuous values เลย

สำหรับ missing values ของสองตัวแปรนั้น จะต้องตัดสินใจจากเหตุการณ์แวดล้อมด้วยว่า ควรแทนค่า missing values อย่างไร

แต่ ณ ที่นี้ จะแทนค่า missing values ด้วย 0 ตามทัศนคติ conservative risk management ที่ว่าควรสมมติให้สิ่งที่ยังไม่ทราบข้อมูลให้มีความเสี่ยงสูงไว้ก่อน:

In [12]:
loan_data_defaults['mths_since_last_delinq'].fillna(0, inplace = True)

In [13]:
loan_data_defaults['mths_since_last_record'].fillna(0, inplace=True)

---
---
---

# Dependent Variables

Column ชื่อ "recoveries" คือ ความเสียหายเมื่อผิดนัดชำระหนี้ที่ได้รับการ recovered (ชำระคืน) แล้ว

สมมติให้ค่า funded_amnt (จำนวนเงินที่ปล่อยกู้) ของผู้ขอกู้ที่ผิดนัดชำระหนี้ เป็นจำนวนความเสียหายรวมเมื่อลูกหนี้ผิดนัดชำระหนี้

ทำการคำนวณตัวแปรตาม ชื่อ recovery rate (ร้อยละความเสียหายที่กู้คืนได้) สำหรับนำไปใช้ใน LGD model:

In [14]:
loan_data_defaults['recovery_rate'] = loan_data_defaults['recoveries'] / loan_data_defaults['funded_amnt']

แสดงสถิติของ column ชื่อ recovery_rate:

In [15]:
loan_data_defaults['recovery_rate'].describe()

count    43236.000000
mean         0.060820
std          0.089770
min          0.000000
25%          0.000000
50%          0.029466
75%          0.114044
max          1.220774
Name: recovery_rate, dtype: float64

จากข้างบนพบว่า สำหรับค่า  recovery rate (ร้อยละความเสียหายที่กู้คืนได้):

- ต่ำสุด คือ 0 ซึ่งมีค่าปกติเป็นไปตามที่ควรจะเป็น

- สูงสุด คือ มากกว่า 1 ซึ่งไม่ปกติ เพราะปกติค่า rate ควรมีค่าระหว่าง 0 ถึง 1

บางครั้งจะพบว่าค่าอยู่นอกพิสัย 0 ถึง 1 อันเนื่องมาจาก การคำนวณทางบัญชีบางอย่าง หรือ ความแตกต่างกันของฐานข้อมูล สำหรับกรณีเหล่านี้ สามารถตัดค่าที่อยู่นอกพิสัย 0 ถึง 1 ให้เท่ากับ 0 หรือ 1 (นั่นคือ มากกว่า 1 ให้เปลี่ยนเป็น 1 และ น้อยกว่า 0 ให้เปลี่ยนเป็น 0)

กระทำได้ดังนี้:

In [16]:
loan_data_defaults['recovery_rate'] = np.where(loan_data_defaults['recovery_rate'] > 1, 1, loan_data_defaults['recovery_rate'])
loan_data_defaults['recovery_rate'] = np.where(loan_data_defaults['recovery_rate'] < 0, 0, loan_data_defaults['recovery_rate'])

แสดงสถิติของ column ชื่อ recovery_rate:

In [17]:
loan_data_defaults['recovery_rate'].describe()

count    43236.000000
mean         0.060770
std          0.089202
min          0.000000
25%          0.000000
50%          0.029466
75%          0.114044
max          1.000000
Name: recovery_rate, dtype: float64

ข้างบนนี้คือตัวแปร recovery rate (ร้อยละความเสียหายที่กู้คืนได้) ซึ่งได้ทำการ preprocessed เสร็จเรียบร้อยแล้ว ตัวแปรตามตัวนี้จะถูกนำไปใช้ใน **LGD model** ต่อไป

---

สำหรับเงินกู้หมุนเวียน เช่น บัตรเครดิต, สินเชื่อบางประเภท, ฯลฯ ผู้ขอกู้สามารถจ่ายหนี้ จากนั้นก็นำเงินในส่วนที่จ่ายไปแล้วออกมาใช้ แล้วก็จ่ายคืนอีก แล้วก็นำเงินออกมาใช้อีก ได้เรื่อย ๆ ตราบใดที่ยังไม่เกินวงเงิน (credit limit)

ดังนั้น ผู้ขอกู้อาจจะจ่ายคืนหนี้บางส่วนไปแล้วเมื่อตอนที่เกิดการผิดนัดชำระหนี้

แปลว่า ผู้ขอกู้ก็จะผิดนัดชำระหนี้แค่เพียงบางส่วน ของจำนวนเงินที่ได้รับอนุมัติตั้งแต่ต้น (จำนวนเงินที่ผิดนัดชำระหนี้แค่เพียงบางส่วน ต่อจำนวนเงินที่ได้รับอนุมัติตั้งแต่ต้น เรียกว่า Credit Conversion Factor)

Exposure at Default (EAD) = Credit Conversion Factor * total funded amount

Credit Conversion Factor (CCF: ค่าแปลงสภาพ) เป็นตัวแปรตาม คำนวณได้ดังนี้:

$$~~~CCF = \frac{funded\_amnt~ - ~total\_rec\_prncp}{funded\_amnt}$$

- funded_amnt (จำนวนเงินที่ปล่อยกู้)
- total_rec_prncp (จำนวนเงินต้นทั้งหมดที่ได้ชำระแล้ว)

ค่า CCF (Credit Conversion Factor) คืออัตราส่วนของเงินต้นที่ยังค้างชำระต่อจำนวนเงินที่ปล่อยกู้ ณ ขณะที่ผู้ขอกู้ผิดนัดชำระหนี้ถ้าหากว่าผู้ขอกู้ผิดนัดชำระหนี้; นั่นคือ หากค่า CCF ยิ่งสูง ยอดหนี้ ณ ขณะที่ผู้ขอกู้ผิดนัดชำระหนี้ยิ่งสูง; นั่นคือ ยิ่งเสี่ยง

คำนวณ Credit Conversion Factor (CCF: ค่าแปลงสภาพ):

In [18]:
loan_data_defaults['CCF'] = (loan_data_defaults['funded_amnt'] - loan_data_defaults['total_rec_prncp']) / loan_data_defaults['funded_amnt']

Credit Conversion Factor (CCF: ค่าแปลงสภาพ)

- CCF = 0 คือ จ่ายครบถ้วนแล้ว
- CCF = 1 คือ ยังไม่ได้จ่ายอะไรเลย

แสดงสถิติของ column ชื่อ CCF (Credit Conversion Factor):

In [19]:
loan_data_defaults['CCF'].describe()

count    43236.000000
mean         0.735952
std          0.200742
min          0.000438
25%          0.632088
50%          0.789908
75%          0.888543
max          1.000000
Name: CCF, dtype: float64

จากข้างบนพบว่า สำหรับค่า CCF นั้น มีค่าต่ำสุดเท่ากับ 0 และสูงสุดเท่ากับ 1 ซึ่งมีค่าปกติเป็นไปตามที่ควรจะเป็น ตัวแปรตามตัวนี้จะถูกนำไปใช้ใน **EAD model** ต่อไป

---

ทำการบันทึก DataFrame ที่ได้คำนวณตัวแปรตามเหล่านี้แล้ว:

- Recovery rate (ร้อยละความเสียหายที่กู้คืนได้) สำหรับนำไปใช้ใน **LGD model**
- Credit Conversion Factor (CCF: ค่าแปลงสภาพ) สำหรับนำไปใช้ใน **EAD model**

ลง disk:

In [20]:
loan_data_defaults.to_csv('./data/loan_data_defaults.csv')

---

# End of notebook.

---